File: src/linear_algebra_linear_independence_easy.py
"""
linear_algebra - linear_independence (easy)
Generated: 2026-02-22T05:57:01.464546
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: Check if 2 vectors in R^2 are linearly independent (visual/simple)
        # ELO ~1000-1150: Direct check with 2 vectors
        a1, a2 = nonzero(-5, 5), nonzero(-5, 5)
        b1, b2 = nonzero(-5, 5), nonzero(-5, 5)
        
        # Make sure they're not scalar multiples
        if a1 * b2 != a2 * b1:
            # Linearly independent
            ans = True
            det_val = a1 * b2 - a2 * b1
            return problem(
                question=f"Are the vectors $\\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}}$ and $\\begin{{pmatrix}} {b1} \\\\ {b2} \\end{{pmatrix}}$ linearly independent?",
                answer=ans,
                difficulty=(1000, 1150),
                topic="linear_algebra/linear_independence",
                solution=steps(
                    f"Two vectors in $\\mathbb{{R}}^2$ are linearly independent if one is not a scalar multiple of the other.",
                    f"Check if $\\begin{{vmatrix}} {a1} & {b1} \\\\ {a2} & {b2} \\end{{vmatrix}} \\neq 0$",
                    f"Determinant = ${a1} \\cdot {b2} - {a2} \\cdot {b1} = {det_val} \\neq 0$",
                    f"Therefore, the vectors are linearly independent.",
                ),
            )
        else:
            # Make them dependent
            k = a1 // b1 if b1 != 0 else a2 // b2
            b1, b2 = a1 // k if k != 0 else a1, a2 // k if k != 0 else a2
            ans = False
            return problem(
                question=f"Are the vectors $\\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}}$ and $\\begin{{pmatrix}} {b1} \\\\ {b2} \\end{{pmatrix}}$ linearly independent?",
                answer=ans,
                difficulty=(1000, 1150),
                topic="linear_algebra/linear_independence",
                solution=steps(
                    f"Two vectors in $\\mathbb{{R}}^2$ are linearly independent if one is not a scalar multiple of the other.",
                    f"Notice that $\\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}} = {k} \\begin{{pmatrix}} {b1} \\\\ {b2} \\end{{pmatrix}}$",
                    f"Therefore, the vectors are linearly dependent (not independent).",
                ),
            )
    
    elif problem_type == 2:
        # Type 2: Check if a single vector is linearly independent (trivial case)
        # ELO ~1000-1100: Understanding that one nonzero vector is always LI
        a1, a2 = nonzero(-5, 5), nonzero(-5, 5)
        ans = True
        return problem(
            question=f"Is the set containing only the vector $\\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}}$ linearly independent?",
            answer=ans,
            difficulty=(1000, 1100),
            topic="linear_algebra/linear_independence",
            solution=steps(
                f"A set containing a single nonzero vector is always linearly independent.",
                f"Since $\\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}} \\neq \\begin{{pmatrix}} 0 \\\\ 0 \\end{{pmatrix}}$, the set is linearly independent.",
            ),
        )
    
    elif problem_type == 3:
        # Type 3: Check if set containing zero vector is LI
        # ELO ~1100-1200: Recognize zero vector makes set dependent
        a1, a2 = nonzero(-5, 5), nonzero(-5, 5)
        ans = False
        return problem(
            question=f"Is the set $\\left\\{{\\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}}, \\begin{{pmatrix}} 0 \\\\ 0 \\end{{pmatrix}}\\right\\}}$ linearly independent?",
            answer=ans,
            difficulty=(1100, 1200),
            topic="linear_algebra/linear_independence",
            solution=steps(
                f"A set of vectors is linearly dependent if it contains the zero vector.",
                f"Since $\\begin{{pmatrix}} 0 \\\\ 0 \\end{{pmatrix}}$ is in the set, we can write:",
                f"$0 \\cdot \\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}} + 1 \\cdot \\begin{{pmatrix}} 0 \\\\ 0 \\end{{pmatrix}} = \\begin{{pmatrix}} 0 \\\\ 0 \\end{{pmatrix}}$ (nontrivial linear combination)",
                f"Therefore, the set is linearly dependent (not independent).",
            ),
        )
    
    else:
        # Type 4: Simple 2x2 determinant check with nice numbers
        # ELO ~1150-1300: Computing determinant to check independence
        a1, a2 = randint(1, 4), randint(1, 4)
        b1, b2 = randint(1, 4), randint(1, 4)
        
        det_val = a1 * b2 - a2 * b1
        
        if det_val != 0:
            ans = True
            return problem(
                question=f"Determine whether the vectors $\\mathbf{{v}}_1 = \\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}}$ and $\\mathbf{{v}}_2 = \\begin{{pmatrix}} {b1} \\\\ {b2} \\end{{pmatrix}}$ are linearly independent.",
                answer=ans,
                difficulty=(1150, 1300),
                topic="linear_algebra/linear_independence",
                solution=steps(
                    f"Form the matrix $A = \\begin{{pmatrix}} {a1} & {b1} \\\\ {a2} & {b2} \\end{{pmatrix}}$ with the vectors as columns.",
                    f"Compute $\\det(A) = {a1} \\cdot {b2} - {a2} \\cdot {b1} = {a1 * b2} - {a2 * b1} = {det_val}$",
                    f"Since $\\det(A) \\neq 0$, the vectors are linearly independent.",
                ),
            )
        else:
            ans = False
            return problem(
                question=f"Determine whether the vectors $\\mathbf{{v}}_1 = \\begin{{pmatrix}} {a1} \\\\ {a2} \\end{{pmatrix}}$ and $\\mathbf{{v}}_2 = \\begin{{pmatrix}} {b1} \\\\ {b2} \\end{{pmatrix}}$ are linearly independent.",
                answer=ans,
                difficulty=(1150, 1300),
                topic="linear_algebra/linear_independence",
                solution=steps(
                    f"Form the matrix $A = \\begin{{pmatrix}} {a1} & {b1} \\\\ {a2} & {b2} \\end{{pmatrix}}$ with the vectors as columns.",
                    f"Compute $\\det(A) = {a1} \\cdot {b2} - {a2} \\cdot {b1} = {a1 * b2} - {a2 * b1} = {det_val}$",
                    f"Since $\\det(A) = 0$, the vectors are linearly dependent (not independent).",
                ),
            )

emit(generate())

File: src/linear_algebra_linear_independence_easy.jl
# linear_algebra - linear_independence (easy)
# Generated: 2026-03-08T21:12:07.337022

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("linear_algebra/linear_independence")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Type 1: Check if two 2D vectors are linearly independent (ELO 700-900)
        a1 = nonzero(-8, 8)
        a2 = nonzero(-8, 8)
        b1 = nonzero(-8, 8)
        b2 = nonzero(-8, 8)
        
        # Ensure they're not scalar multiples
        while abs(a1 * b2 - a2 * b1) < 2
            b1 = nonzero(-8, 8)
            b2 = nonzero(-8, 8)
        end
        
        det_val = a1 * b2 - a2 * b1
        is_indep = det_val != 0
        
        v1_str = "\\begin{bmatrix} $(a1) \\\\ $(a2) \\end{bmatrix}"
        v2_str = "\\begin{bmatrix} $(b1) \\\\ $(b2) \\end{bmatrix}"
        
        problem(
            question="Are the vectors \$$(v1_str)\$ and \$$(v2_str)\$ linearly independent? Answer 'yes' or 'no'.",
            answer=is_indep ? "yes" : "no",
            difficulty=(700, 900),
            answer_type="text",
            solution=steps(
                "Two vectors in ℝ² are linearly independent if and only if they are not scalar multiples of each other.",
                "Check the determinant: \$$(a1) \\cdot $(b2) - $(a2) \\cdot $(b1) = $(det_val)\$",
                det_val != 0 ? "Since the determinant is nonzero, the vectors are linearly independent." : "Since the determinant is zero, the vectors are linearly dependent.",
                sol("Answer", is_indep ? "yes" : "no")
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Type 2: Determine if three 2D vectors are linearly independent (ELO 800-1000)
        a1 = nonzero(-7, 7)
        a2 = nonzero(-7, 7)
        b1 = nonzero(-7, 7)
        b2 = nonzero(-7, 7)
        c1 = nonzero(-7, 7)
        c2 = nonzero(-7, 7)
        
        v1_str = "\\begin{bmatrix} $(a1) \\\\ $(a2) \\end{bmatrix}"
        v2_str = "\\begin{bmatrix} $(b1) \\\\ $(b2) \\end{bmatrix}"
        v3_str = "\\begin{bmatrix} $(c1) \\\\ $(c2) \\end{bmatrix}"
        
        problem(
            question="Are the vectors \$$(v1_str)\$, \$$(v2_str)\$, and \$$(v3_str)\$ linearly independent? Answer 'yes' or 'no'.",
            answer="no",
            difficulty=(800, 1000),
            answer_type="text",
            solution=steps(
                "We have three vectors in ℝ², which is a 2-dimensional space.",
                "Any set of more than 2 vectors in ℝ² must be linearly dependent.",
                "This is because the maximum number of linearly independent vectors in ℝⁿ is n.",
                sol("Answer", "no")
            ),
            time=75
        )
        
    elseif problem_type == 3
        # Type 3: Check if two 3D vectors are linearly independent (ELO 750-950)
        a1 = nonzero(-6, 6)
        a2 = nonzero(-6, 6)
        a3 = nonzero(-6, 6)
        
        # Create second vector not a scalar multiple
        b1 = nonzero(-6, 6)
        b2 = nonzero(-6, 6)
        b3 = nonzero(-6, 6)
        
        # Check not parallel (at least one 2x2 minor is nonzero)
        while abs(a1*b2 - a2*b1) < 1 && abs(a1*b3 - a3*b1) < 1 && abs(a2*b3 - a3*b2) < 1
            b1 = nonzero(-6, 6)
            b2 = nonzero(-6, 6)
            b3 = nonzero(-6, 6)
        end
        
        v1_str = "\\begin{bmatrix} $(a1) \\\\ $(a2) \\\\ $(a3) \\end{bmatrix}"
        v2_str = "\\begin{bmatrix} $(b1) \\\\ $(b2) \\\\ $(b3) \\end{bmatrix}"
        
        problem(
            question="Are the vectors \$$(v1_str)\$ and \$$(v2_str)\$ linearly independent? Answer 'yes' or 'no'.",
            answer="yes",
            difficulty=(750, 950),
            answer_type="text",
            solution=steps(
                "Two vectors are linearly independent if neither is a scalar multiple of the other.",
                "Check if \$\\mathbf{v}_2 = k\\mathbf{v}_1\$ for some scalar k.",
                "If this were true, we'd need \$$(b1) = k \\cdot $(a1)\$, \$$(b2) = k \\cdot $(a2)\$, and \$$(b3) = k \\cdot $(a3)\$ for the same k.",
                "These ratios are not all equal, so the vectors are linearly independent.",
                sol("Answer", "yes")
            ),
            time=100
        )
        
    else
        # Type 4: Find value that makes vectors linearly dependent (ELO 900-1200)
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = nonzero(-5, 5)
        d = nonzero(-5, 5)
        
        # Second vector with parameter k
        # Want det = 0, so a*k - b*c = 0, thus k = bc/a
        k_answer = (b * c) // a
        
        v1_str = "\\begin{bmatrix} $(a) \\\\ $(b) \\end{bmatrix}"
        v2_str = "\\begin{bmatrix} k \\\\ $(c) \\end{bmatrix}"
        
        problem(
            question="Find the value of k that makes the vectors \$$(v1_str)\$ and \$$(v2_str)\$ linearly dependent.",
            answer=k_answer,
            difficulty=(900, 1200),
            solution=steps(
                "Two vectors are linearly dependent if one is a scalar multiple of the other, or equivalently, if their determinant is zero.",
                sol("Determinant", "$(a) \\cdot $(c) - $(b) \\cdot k = 0"),
                "Solve for k: \$$(a * c) - $(b)k = 0\$",
                sol("k", "\\frac{$(a * c)}{$(b)} = $(k_answer)")
            ),
            time=120
        )
    end
end